pkg/kine/server/watch.go

```diff
@@ -4,49 +4,78 @@ import (
 	"context"
 	"sync"
 	"sync/atomic"
+	"time"
 
 	"github.com/sirupsen/logrus"
 	"go.etcd.io/etcd/api/v3/etcdserverpb"
 	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/exp/rand"
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 var (
 	watchID int64
 )
 
+// explicit interface check
+var _ etcdserverpb.WatchServer = (*KVServerBridge)(nil)
+
+// getProgressReportInterval returns the configured progress report interval, with some jitter
+func (s *KVServerBridge) getProgressReportInterval() time.Duration {
+	// add rand(1/10*notifyInterval) as jitter so that kine will not
+	// send progress notifications to watchers at the same time even when watchers
+	// are created at the same time.
+	jitter := time.Duration(rand.Int63n(int64(s.limited.notifyInterval) / 10))
+	return s.limited.notifyInterval + jitter
+}
+
 func (s *KVServerBridge) Watch(ws etcdserverpb.Watch_WatchServer) error {
 	w := watcher{
-		server:  ws,
-		backend: s.limited.backend,
-		watches: map[int64]func(){},
+		server:   ws,
+		backend:  s.limited.backend,
+		watches:  map[int64]func(){},
+		progress: map[int64]chan<- int64{},
 	}
 	defer w.Close()
 
+	go wait.PollInfiniteWithContext(ws.Context(), s.getProgressReportInterval(), w.ProgressIfSynced)
```

Comment:
The method is deprecated, but in general we are adding a rather big dependency for a very simple functionality.

Can't we just write a polling loop here? I think if would also be cleaner.

Reply:
Hm, yes it's a shortcut. The poll loop simplifies things nicely.